Pegasus/include/BoundingVolumes.hpp

```diff
@@ -133,6 +140,362 @@ class BoundingSphere
     );
 };
 } // namespace sphere
+
+namespace hierarchy
+{
+
+struct SplitIndices
+{
+    Indices lowerIndices;
+    Indices upperIndices;
+};
+
+template <typename BoundingVolume>
+struct SimpleShapeExtractor
+{
+};
+
+template <>
+struct SimpleShapeExtractor<aabb::AxisAlignedBoundingBox>
+{
+    Box GetSimpleShape(aabb::AxisAlignedBoundingBox const& aabb)
+    {
+        return aabb.GetBox();
+    }
+};
+
+template <>
+struct SimpleShapeExtractor<obb::OrientedBoundingBox>
+{
+    Box GetSimpleShape(obb::OrientedBoundingBox const& obb)
+    {
+        return obb.GetBox();
+    }
+};
+
+template <>
+struct SimpleShapeExtractor<sphere::BoundingSphere>
+{
+    Sphere GetSimpleShape(sphere::BoundingSphere const& sphere)
+    {
+        return sphere.GetSphere();
+    }
+};
+
+template <typename BoundingVolume>
+class BoundingVolumeHierarchy
+{
+public:
+    class Node
+    {
+    public:
+        BoundingVolume const volume;
+
+        Node(Node const&) = delete;
+        Node(Node &&) = delete;
+        Node & operator=(Node const&) = delete;
+        Node & operator=(Node &&) = delete;
+
+        std::unique_ptr<Node> const& GetLowerChild() const
+        {
+            return lowerChild;
+        }
+
+        std::unique_ptr<Node> const& GetUpperChild() const
+        {
+            return upperChild;
+        }
+
+        bool IsLeaf() const
+        {
+            return isLeaf;
+        }
+
+        ~Node() = default;
+
+    private:
+        mutable std::unique_ptr<Node> lowerChild;
+        mutable std::unique_ptr<Node> upperChild;
+        bool isLeaf;
+
+        explicit Node(BoundingVolume && volume)
+            : volume(volume)
+            , lowerChild(nullptr)
+            , upperChild(nullptr)
+            , isLeaf(true)
+        {}
+
+        friend class BoundingVolumeHierarchy;
+    };
+
+private:
+    using CentroidsMap = std::unordered_map<std::size_t, glm::dvec3>;
+    using NodePtr = std::unique_ptr<Node>;
+
+    std::size_t static const MAX_NODE_SIZE = 5;
+
+    Shape const& shape;
+    NodePtr root;
+    mutable SimpleShapeIntersectionDetector detector;
+    mutable SimpleShapeExtractor<BoundingVolume> extractor;
+
+    Vertices GetShapeVertices(Indices const& indices) const
+    {
+        Vertices result;
+        result.reserve(3 * indices.size());
```

Comment:
mmm seems like indices size() can be half of result size, that's mean that reserve will throw exception -> we haven't exceptions->crash. 
May be check for max_count ?

Reply:
Why can it be half of `result` size? `result` size directly depends on `indices` size.